src/backend/opencl/program.cpp

```diff
@@ -67,4 +69,32 @@ namespace opencl
             throw;
         }
     }
+
+    NDRange getKernelLaunchCfg1D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = getDeviceWorkGroupSize(kernel, device);
+
+        // make sure you are not launching what the device can't handle
+        dim0ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim0ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, 1, 1);
+    }
+
+    NDRange getKernelLaunchCfg2D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = kernel.getWorkGroupInfo<CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE>(getDevice(device));
+
+        size_t recomWorkGroupSize = getDeviceWorkGroupSize(kernel, device);
+
+        size_t dim1ThreadCount = divup(recomWorkGroupSize, dim0ThreadCount);
+
+        // make sure you are not launching what the device can't handle
+        dim1ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim1ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, dim1ThreadCount, 1);
+    }
```

Comment:
I agree with @pavanky. I don't think we can use this function in it's current state. We need to have a function LIKE this but not in its current implementation.

Many of the kernels we are using have hard coded values they expect for the NDRange. Of course this is not ideal because each device is different. This function should have a lookup table that takes into account the device and the kernel to return an appropriate value.

An even better solution would return a kernel and its configuration given a generic name for the kernel and its device. If there are no specializations for a specific kernel a default version would be returned.